### PR TITLE
Run acceptance tests for the finance reports

### PR DESCRIPTION
**Pull request by mulby**
This includes:
- Framework fixes to allow this to run on devstack
- Removal of database imports that we didn't end up actually using

@brianhw - This is ready for a first pass of review.

FYI - @jab5569

### CODE REVIEW

File: edx/analytics/tasks/reports/orders_import.py [hunk 1]
```diff
@@ -190,6 +108,11 @@ def columns(self):
     def partition(self):
         return HivePartition('dt', self.import_date.isoformat())  # pylint: disable=no-member
 
+    def hiveconfs(self):
```
Comment (mulby) [inline edx/analytics/tasks/reports/orders_import.py:113]: This is the hack that is required to make the test pass, but really shouldn't be necessary.

File: edx/analytics/tasks/tests/acceptance/services/vertica.py [hunk 2]
```diff
@@ -58,11 +62,16 @@ def connect(self):
         """
         Connect to the Vertica server.
         """
+        if self.disabled:
+            raise RuntimeError('The vertica service is disabled')
```
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py]: This will cause tests that require vertica to fail when credentials haven't been specified. We may want to cause them to be skipped instead.

File: edx/analytics/tasks/tests/acceptance/__init__.py [hunk 3]
```diff
@@ -74,8 +72,9 @@ def setUp(self):
         self.catalog_path = 'http://acceptance.test/api/courses/v2'
         database_name = 'test_' + self.identifier
         schema = 'test_' + self.identifier
-        import_database_name = 'import_' + database_name
-        export_database_name = 'export_' + database_name
+        import_database_name = 'reports_import_' + database_name
+        export_database_name = 'reports_export_' + database_name
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/__init__.py]: Our nomenclature for import vs. export is not very consistent in general, but this is confusing to me.  Generally, when we talk about imports, we mean loading data into reports (or resultstore).  I thought that when we used 'export' we meant exports mostly from LMS DB.  New additions start to challenge these assumptions:  reading from APIs for course subjects is one example of a different export source.  And we load data into Vertica as well as into the resultstore.  

That said, my issue here is that I don't know what the "reports_" prefix is meaning.  The import is presumably from LMS for this, and has nothing to do with the resultstore.  But maybe "reports_"  simply means that it's a separate database used only for generating reports?

File: edx/analytics/tasks/tests/acceptance/__init__.py [hunk 4]
```diff
@@ -74,8 +72,9 @@ def setUp(self):
...
+        otto_database_name = 'reports_otto_' + database_name
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/__init__.py]: Oh, similar question about this - what does "reports_" mean?

File: edx/analytics/tasks/tests/acceptance/test_financial_reports.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    "order_refunded_quantity",
+    "order_user_id",
+    "order_username",
+    "order_user_email",
+    "order_product_class",
+    "order_product_detail",
+    "order_course_id",
+    "order_org_id",
+    "order_processor",
+]
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/test_financial_reports.py]: Does it add value to hard-code this here?  I'm wondering if it makes more sense to determine this list dynamically, rather than cut-paste maintenance.  That is, instantiate a LoadInternalReportingOrderTransactionsToWarehouse object, fetch its columns property, and pull the row names out of that.  We're already checking the expected columns implicitly by maintaining an expected output CSV.

File: edx/analytics/tasks/reports/orders_import.py [hunk 6]
```diff
@@ -99,63 +55,25 @@ def requires(self):
...
-                **kwargs
-            )
+            # Original shopping cart tables.
+            ImportShoppingCartOrder(**kwargs),
+            ImportShoppingCartOrderItem(**kwargs),
+            ImportShoppingCartCertificateItem(**kwargs),
+            ImportShoppingCartPaidCourseRegistration(**kwargs),
+            ImportShoppingCartDonation(**kwargs),
+            ImportShoppingCartCourseRegistrationCodeItem(**kwargs),
+            ImportAuthUserTask(**kwargs),
```
Comment (brianhw) [inline edx/analytics/tasks/reports/orders_import.py:78]: Nit:  add comment to indicate "Other LMS tables" (or similar), since this last one is not a shopping cart table.  I would also remove "Original" from the comment above.

File: edx/analytics/tasks/reports/orders_import.py (hunk 1 above)
Comment (brianhw) [inline edx/analytics/tasks/reports/orders_import.py:113]: There should definitely be some kind of comment here explaining the issue as best as possible.  I would have no idea looking at this code to know why this configuration parameter is being set, and why here.

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+--
+
+DROP TABLE IF EXISTS `shoppingcart_courseregcodeitem`;
+CREATE TABLE `shoppingcart_courseregcodeitem` (
+  `orderitem_ptr_id` int(11) NOT NULL,
+  `course_id` varchar(128) NOT NULL,
+  `mode` varchar(50) NOT NULL,
+  PRIMARY KEY (`orderitem_ptr_id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql:12]: Should these have some data too?

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_donation.sql [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+--
+
+DROP TABLE IF EXISTS `shoppingcart_donation`;
+CREATE TABLE `shoppingcart_donation` (
+  `orderitem_ptr_id` int(11) NOT NULL,
+  `donation_type` varchar(32) NOT NULL,
+  `course_id` varchar(255) NOT NULL,
+  PRIMARY KEY (`orderitem_ptr_id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_donation.sql:12]: Should this have some data too?

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_paidcourseregistration.sql [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+DROP TABLE IF EXISTS `shoppingcart_paidcourseregistration`;
+CREATE TABLE `shoppingcart_paidcourseregistration` (
+  `orderitem_ptr_id` int(11) NOT NULL,
+  `course_id` varchar(128) NOT NULL,
+  `mode` varchar(50) NOT NULL,
+  `course_enrollment_id` int(11),
+  PRIMARY KEY (`orderitem_ptr_id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_paidcourseregistration.sql:13]: Should this also have some data?

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/otto/catalogue_productattribute.sql [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+CREATE TABLE `catalogue_productattribute` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `name` varchar(128) NOT NULL,
+  `code` varchar(128) NOT NULL,
+  `type` varchar(20) NOT NULL,
+  `required` tinyint(1) NOT NULL,
+  `option_group_id` int(11) DEFAULT NULL,
+  `product_class_id` int(11),
+  PRIMARY KEY (`id`)
+) ENGINE=InnoDB AUTO_INCREMENT=6 DEFAULT CHARSET=utf8;
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/otto/catalogue_productattribute.sql]: nit: none of the others specify an auto_increment value....

File: edx/analytics/tasks/tests/acceptance/services/vertica.py (hunk 2 above)
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py]: Skipping particular acceptance tests when needed infrastructure is not present seems to be a more reasonable way to allow for open development while still allowing our private add-ons, like exporter and vertica.  For the exporter, though, we annotated the particular test:      @unittest.skipIf(os.getenv('EXPORTER') is None, "requires private Exporter code")

So instead of failing, I think that one can: "raise unittest.SkipTest(reason)".  

Is the configuration of this url in acceptance tests a reasonable way to determine if tests using Vertica?  I don't really remember where this config value comes from.  It may make more sense for each test to determine whether to skip or change its scope of testing based on whether the service is disabled, rather than making it a global here.  For example, the end-to-end test could check if vertica were enabled, and if not, could run a different task and check the reconciliation table instead.  (Note that the output should be close to the same, if not identical.)

File: edx/analytics/tasks/tests/acceptance/services/vertica.py [hunk 11]
```diff
@@ -13,7 +13,11 @@ class VerticaService(object):
     """Service object to be used as a member of a class to enable that class to write to and read from Vertica."""
 
     def __init__(self, config, schema_name):
-        self.vertica_creds_url = config['vertica_creds_url']
+        self.vertica_creds_url = config.get('vertica_creds_url', '')
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py]: nit: is there a reason to use '' and len() here?  Instead of just doing self.blah = get('blah') and if self.blah?  Just curious...

File: edx/analytics/tasks/tests/acceptance/test_financial_reports.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    def prepare_database(self, name, database):
+        sql_fixture_base_url = url_path_join(self.data_dir, 'input', 'finance_reports', name)
+        for filename in os.listdir(sql_fixture_base_url):
+            self.execute_sql_fixture_file(url_path_join(sql_fixture_base_url, filename), database=database)
+
+    def test_end_to_end(self):
+        self.task.launch([
+            'BuildFinancialReportsTask',
+            '--import-date', '2015-09-02',
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/test_financial_reports.py]: Is there a reason why this import date is different than IMPORT_DATE?  The difference is confusing.  Maybe rename IMPORT_DATE if it's not accurate...

**Comment by brianhw** [comment]
@mulby Finished a pass.  Very close!

File: edx/analytics/tasks/reports/orders_import.py (hunk 1 above)
Comment (mulby) [inline edx/analytics/tasks/reports/orders_import.py:113]: Done.

File: edx/analytics/tasks/tests/acceptance/services/vertica.py (hunk 2 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py]: Fixed the vertica case, we'll have to address the exporter case later...

File: edx/analytics/tasks/tests/acceptance/__init__.py (hunk 3 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/__init__.py]: reports_ was an artifact of devstack permissions, I've changed it to "acceptance" to be more clear.

Agreed about import vs. export. They are relative terms and we don't use them consistently.

File: edx/analytics/tasks/tests/acceptance/__init__.py (hunk 4 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/__init__.py]: Done.

File: edx/analytics/tasks/tests/acceptance/test_financial_reports.py (hunk 5 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/test_financial_reports.py]: That's fair... fixing.

File: edx/analytics/tasks/reports/orders_import.py (hunk 6 above)
Comment (mulby) [inline edx/analytics/tasks/reports/orders_import.py:78]: Done

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql (hunk 7 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql:12]: Nope - a given item should appear in exactly one of these tables depending on its type (certificate, coupon etc).

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/otto/catalogue_productattribute.sql (hunk 10 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/otto/catalogue_productattribute.sql]: Oops, good catch. Fixed.

File: edx/analytics/tasks/tests/acceptance/services/vertica.py (hunk 11 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py]: Not a good one. Changing.

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql (hunk 7 above)
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql:12]: Right, but should our end-to-end include a donation, a coupon, etc. for fuller integration coverage?

File: edx/analytics/tasks/tests/acceptance/test_financial_reports.py (hunk 12 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/test_financial_reports.py]: Yeah, I've clarified that. One is an open upper bound, the other is the actual day that contains real data.

File: edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql (hunk 7 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/fixtures/input/finance_reports/lms/shoppingcart_courseregcodeitem.sql:12]: Hrm... that's a reasonable point given that we don't have unit test coverage of these nuances. I'm a little nervous that it could explode the complexity of the testing dataset dramatically if we tried to cover all of the edge cases, though.

Open to your thoughts here...

**Comment by mulby** [comment]
@brianhw addressed all of your comments, running a full acceptance test suite now

File: edx/analytics/tasks/tests/acceptance/services/vertica.py [hunk 13]
```diff
@@ -7,13 +7,18 @@
...
 from edx.analytics.tasks.url import get_target_from_url
+from edx.analytics.tasks.tests import unittest
 
 
 class VerticaService(object):
     """Service object to be used as a member of a class to enable that class to write to and read from Vertica."""
 
     def __init__(self, config, schema_name):
-        self.vertica_creds_url = config['vertica_creds_url']
+        self.vertica_creds_url = config.get('vertica_creds_url')
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py:17]: Just a small source of confusion:  It seems that the setUp method for the AcceptanceTestCase (in **init**) requires that a value be set for vertica_creds_url.  So it must be set there, but presumably must be set to a zero-length string.   Does this actually work?  Or do we have to do a strip on it as well?  Or is it something that we want to make optional at the **init** level?  It seems kind of weird to require a value exist but be empty.

File: edx/analytics/tasks/tests/acceptance/services/vertica.py [hunk 14]
```diff
@@ -58,11 +63,16 @@ def connect(self):
         """
         Connect to the Vertica server.
         """
+        if self.disabled:
+            raise unittest.SkipTest('The vertica service is disabled')
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py:67]: Will this actually mark the test to be skipped in time?  It seems that the time that the VerticaService is actually connected to is when validating the result of some Vertica load, so I would expect the test to have already failed by the time we get here (by trying to actually load data into a non-existent Vertica server).  It may be necessary instead to move this call into each individual test that would _eventually_ use VerticaService.  So it would have code if self.vertica.disabled: raise unittest.SkipTest(...).

**Comment by brianhw** [comment]
Just two more questions.

File: edx/analytics/tasks/tests/acceptance/services/vertica.py (hunk 13 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py:17]: Yeah, if you leave it empty it does what you expect. I like the idea of making it optional, though, I'll make that change.

File: edx/analytics/tasks/tests/acceptance/services/vertica.py (hunk 14 above)
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/services/vertica.py:67]: Not really. It runs the whole test and then skips at the very end when it goes to connect to vertica.

In the future I'd like to add a better pattern that lets you tell the framework whether you are going to use optional features or not (vertica, exporter etc).

I was planning on addressing that in a follow-on story, though, when we solve the devstack more comprehensively.

My intent with this PR is to push the effort forward (largely because it made my development faster), not necessary complete it.

File: edx/analytics/tasks/tests/acceptance/__init__.py [hunk 15]
```diff
@@ -98,6 +98,10 @@ def setUp(self):
                 'credentials': self.config['credentials_file_url'],
                 'database': export_database_name
             },
+            'otto-database-import': {
+                'credentials': self.config['credentials_file_url'],
+                'database': otto_database_name
+            },
             'vertica-export': {
                 'credentials': self.config['vertica_creds_url'],
```
Comment (brianhw) [inline edx/analytics/tasks/tests/acceptance/__init__.py]: If you make it optional for vertica_creds_url to be in config, then this can break ....

Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/__init__.py]: Good catch, fixed.

File: edx/analytics/tasks/tests/acceptance/__init__.py [hunk 16]
```diff
@@ -123,16 +129,21 @@ def setUp(self):
...
         self.hive = hive.HiveService(self.task, self.config, database_name)
+        self.vertica = vertica.VerticaService(self.config, schema)
 
-        self.reset_external_state()
+        if os.getenv('DISABLE_RESET_STATE', 'false').lower() != 'true':
+            self.reset_external_state()
 
     def reset_external_state(self):
-        self.s3_client.remove(self.test_root, recursive=True)
+        root_target = get_target_from_url(self.test_root.replace('s3://', 's3+https://'))
```
Comment (mulby) [inline edx/analytics/tasks/tests/acceptance/__init__.py:148]: This is an awful hack - suggestions for improvement are welcome!

On the jenkins box it doesn't have the hadoop client installed, so it can't process "s3://" urls. However, on the devstack it can't use "s3+https" urls since it can't access S3.

At a minimum I suppose I should include a comment about this.

**Comment by brianhw** [comment]
LGTM for the scope of this story. :+1:
